Sprint 7 - Added the FileBackedTaskManager class, initial commit.

vil-shavetail opened the pull request.

vil-shavetail committed:
Sprint 7 - Added the FileBackedTaskManager class, initial commit.

vil-shavetail committed:
Sprint 7 - Added methods: save(), loadFromFile, historyFromString(), historyToString for the FileBackedTaskManager class. Added tests for new functionality. Added the ManagerSaveException class.

vil-shavetail committed:
Sprint 7 - Fixed checkStyle errors.

vil-shavetail committed:
Sprint 7 - Fixed checkStyle errors.

vil-shavetail committed:
Sprint 7 - Fixed checkStyle errors.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+
+    public FileBackedTaskManager(HistoryManager historyManager) {
+        this.historyManager = historyManager;
+
```
Лишний перенос строки)

vil-shavetail replied:
Убрал

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    public void save() {
+        Path file = Path.of("tasks.csv");
```
Здесь надо использовать файл, который передается в конструкторе при создании менеджера. Добавьте поле в данный класс для хранения файла

vil-shavetail replied:
Исправлено.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+import java.util.Objects;
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+    private final HistoryManager historyManager;
```
Данное поле должно присутствовать только в `InMemoryTaskManager`, так как поле наследуется от `InMemoryTaskManager`

vil-shavetail replied:
Исправлено.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+        }
+        try (FileWriter fileWriter = new FileWriter(String.valueOf(file), StandardCharsets.UTF_8);
+             BufferedWriter bufferedWriter = new BufferedWriter(fileWriter)) {
+            bufferedWriter.write("id,type,name,status,description,epic");
```
Данная строка будет слабо изменяться, по этой причине рекомендую вынести ее в константу

vil-shavetail replied:
Если речь о строке "id,type,name,status,description,epic", то вынес в константу.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+                if (line.isEmpty()) {
+                    break;
+                }
+                if (!line.equals("id,type,name,status,description,epic")) {
```
Здесь так же можно использовать хедер, объявленный выше

vil-shavetail replied:
Исправлено.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    public void loadFromFile(File file) {
```
Согласно ТЗ, данный метод должен быть объявлен, как `static FileBackedTaskManager loadFromFile(File file)`

vil-shavetail replied:
Исправлено

KirPyt reviewed (commented):
Здравствуйте!

Указал некоторые комментарии по коду, которые надо учесть. В частности, стоит добавить поле для хранения резервного файла в `FileBackedTaskManager`, а так же получить менеджер истории за счет наследования

vil-shavetail committed:
Sprint 7 - Fixed the reviewer's remarks after first review.
Fixed:
1) https://github.com/vil-shavetail/java-kanban/pull/2#discussion_r1935694642.

vil-shavetail committed:
Sprint 7 - Fixed the reviewer's remarks after first review.

vil-shavetail committed:
Sprint 7 - Fixed checkStyle errors.

vil-shavetail committed:
Sprint 7 - Fixed checkStyle errors.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+
+    private static HistoryManager historyManager;
```
Данное поле наследуем от `InMemoryTaskManager`, здесь его не нужно создавать отдельно

vil-shavetail replied:
Исправил

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+    private final File file;
+    private static final String HEADER = "id,type,name,status,description,epic";
+
+    public FileBackedTaskManager(HistoryManager historyManager, File file) {
```
В конструктор передаем только файл, менеджер истории создается через утилитарный класс `Managers` при создании объекта `FileBackedTaskManager` как наследника `InMemoryTaskManager`

vil-shavetail replied:
Исправил вызов в конструкторе - super(Managers.getDefaultHistory());

KirPyt reviewed (commented):
Здравствуйте!

Указал два замечания по коду, которые надо поправить, чтобы выстроить корректную структуру работы с менеджером истории

vil-shavetail committed:
Sprint 7 - Fixed the reviewer's remarks after second review.

KirPyt commented on src/ru/yandex/practicum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,255 @@
+    private static final String HEADER = "id,type,name,status,description,epic";
+
+    public FileBackedTaskManager(File file) {
+        super(Managers.getDefaultHistory());
```
Можно сразу явным образом описать инициализацию менеджера истории, так как других реализаций менеджера истории нет, то можно явно его инициализировать в `InMemoryTaskManager`, чтобы не передавать как отдельный параметр

KirPyt reviewed (commented):
Здравствуйте!

В целом критических замечаний больше нет, Ваш проект зачтен, хорошего дня!
